models/pathology_nuclei_segmentation_classification/scripts/prepare_patches.py

```diff
@@ -0,0 +1,219 @@
+            step_size : a tuple of (h, w)
+        Return:
+            a list of sub patches, each patch is same dtype as x
+
+        """
+        diff_h = self.patch_size[0] - self.step_size[0]
+        padt = diff_h // 2
+        padb = diff_h - padt
+
+        diff_w = self.patch_size[1] - self.step_size[1]
+        padl = diff_w // 2
+        padr = diff_w - padl
+
+        pad_type = "reflect"
+        x = np.lib.pad(x, ((padt, padb), (padl, padr), (0, 0)), pad_type)
+        sub_patches = self.__extract_valid(x)
+        return sub_patches
+
+    def extract(self, x, patch_type):
+        patch_type = patch_type.lower()
+        self.patch_type = patch_type
+        if patch_type == "valid":
+            return self.__extract_valid(x)
+        elif patch_type == "mirror":
+            return self.__extract_mirror(x)
+        else:
+            raise ValueError(f"Unknown Patch Type {patch_type}")
+
+
+def main(cfg):
+    xtractor = PatchExtractor(cfg["patch_size"], cfg["step_size"])
+
+    for phase in ["Train", "Test"]:
+        img_dir = os.path.join(cfg["root"], f"{phase}/Images")
+        ann_dir = os.path.join(cfg["root"], f"{phase}/Labels")
+
+        file_list = glob.glob(os.path.join(ann_dir, "*mat"))
+        file_list.sort()  # ensure same ordering across platform
+
+        out_dir = f"{cfg['root']}/Prepared/{phase}"
+        if os.path.isdir(out_dir):
+            shutil.rmtree(out_dir)
+        os.makedirs(out_dir)
+
+        pbar_format = "Process File: |{bar}| {n_fmt}/{total_fmt}[{elapsed}<{remaining},{rate_fmt}]"
+        pbarx = tqdm.tqdm(total=len(file_list), bar_format=pbar_format, ascii=True, position=0)
+
+        for file_path in file_list:
+            base_name = pathlib.Path(file_path).stem
+
+            img = load_img(f"{img_dir}/{base_name}.png")
+            ann = load_ann(f"{ann_dir}/{base_name}.mat")
+
+            # *
+            img = np.concatenate([img, ann], axis=-1)
+            sub_patches = xtractor.extract(img, cfg["extract_type"])
+
+            pbar_format = "Extracting  : |{bar}| {n_fmt}/{total_fmt}[{elapsed}<{remaining},{rate_fmt}]"
+            pbar = tqdm.tqdm(total=len(sub_patches), leave=False, bar_format=pbar_format, ascii=True, position=1)
+
+            for idx, patch in enumerate(sub_patches):
+                image_patch = patch[..., :3].transpose(2, 0, 1)  # make channel first
+                inst_map_patch = patch[..., 3][None]  # add channel dim
+                type_map_patch = patch[..., 4][None]  # add channel dim
+                np.save("{0}/{1}_{2:03d}_image.npy".format(out_dir, base_name, idx), image_patch)
+                np.save("{0}/{1}_{2:03d}_inst_map.npy".format(out_dir, base_name, idx), inst_map_patch)
+                np.save("{0}/{1}_{2:03d}_type_map.npy".format(out_dir, base_name, idx), type_map_patch)
+                pbar.update()
+            pbar.close()
+            # *
+
+            pbarx.update()
+        pbarx.close()
+
+
+def parse_arguments():
```

Comment:
I have fewer 2000x2000 newly annotated image + label files.. can i use this utility to prepare the required patches to train or fine tune the model?  what is the requirement to create new patches for a given image+label pair?

label => labelmap (mask with class values [0...4])

``` json
{
  "Background": 0,
  "Other": 1,
  "Inflammatory": 2,
  "Epithelial": 3,
  "Spindle-Shaped": 4
}
```

Reply:
Hi @SachidanandAlle, I think nothing special should be considered, only this `load_ann` function, which is specific to preparing the CoNSeP dataset. 
https://github.com/Project-MONAI/model-zoo/blob/3bff1f2de0ed7336eb627c16f04663105e39bde7/models/pathology_nuclei_segmentation_classification/scripts/prepare_patches.py#L18

And also this part, If you think we need to be compatible with some of the user's own data such as the data path, maybe I can make it a little more robust. 
https://github.com/Project-MONAI/model-zoo/blob/3bff1f2de0ed7336eb627c16f04663105e39bde7/models/pathology_nuclei_segmentation_classification/scripts/prepare_patches.py#L155-L174

Reply:
Hi @SachidanandAlle, what do you think about it?

Reply:
👍

Reply:
Hi @SachidanandAlle, could you please help review it again, thanks!